core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -619,6 +619,9 @@ private[deploy] class Master(
         case e: Exception => logInfo("Worker " + worker.id + " had exception on reconnect")
       }
     }
+
+    // In case of zero workers and apps, we can complete recovery.
+    if (canCompleteRecovery) { completeRecovery() }
   }
 
```

Comment:
If we are completing recovery when `canCompleteRecovery == true`, we can avoid scheduling `CompleteRecovery` task ?
Perhaps returns a `boolean` from this method to indicate it ?